Feat/custom cmd

HendriceH opened the pull request:
Added two ways to execute custom commands through `obr submit` 
1. `obr submit -o runParallelSolver --solver-cmd "postProcess -func yplus"` 
2. Prescribing a series of commands inside the YAML file. 
```
pre_cmds:
        - potentialFoam
        - renumberMesh -overwrite
post_cmds: 
        - postProcess -func yPlus
        - postProcess -func Q -latestTime
```
 The functionality is added to read the command from file and execute them with the `runParallelPre` and `runParallelPost` operations. The whole chain can be called via the operations group `simulate`
  `obr submit -o simulate`
  Note following implementation details:
  - The pre and post operation is only executed if commands are present in the YAML
  - Additionally a `tasksPerNode` keyword is added for the decomposePar operation, which can be used via the scheduler templates to submit varying numbers of tasks per node within one study

greole committed:
wip add postprocess

greole committed:
wip add postprocess

greole committed:
wip add postprocess

greole committed:
wip add postprocess

greole committed:
add example

greole committed:
fix force int for subdomain

greole committed:
add preflight example

greole committed:
WIP dump env

greole committed:
fix issue mit multiple custom matcher

greole committed:
fix issue mit multiple custom matcher

greole committed:
fix issue mit multiple custom matcher

greole committed:
fix issue mit multiple custom matcher

HendriceH committed:
Added custom --solver-cmd option to inject custom command with options instead of application read from controlDict

HendriceH committed:
Added custom --solver-cmd option to inject custom command with options instead of application read from controlDict

HendriceH committed:
Multilevel scotch implementation

HendriceH committed:
Merge branch 'feat/custom_cmd' into enh/exasim_meeting

HendriceH committed:
Merge pull request #212 from exasim-project/enh/exasim_meeting

Enh/exasim meeting

HendriceH committed:
Added custom pre and post-processing commands. They can be bundled in one slurm execution with 'simulate' group operation

HendriceH committed:
Added custom pre and post-processing commands. They can be bundled in one slurm execution with 'simulate' group operation

HendriceH committed:
Merged PR213 into PR214

HendriceH committed:
Added tasksPerNode keyword for decomposeParDict to be calculated in slurm template to run with varying ranksPerGPU

HendriceH committed:
Added tasksPerNode keyword for decomposeParDict to be calculated in slurm template to run with varying ranksPerGPU

HendriceH committed:
Fixed runParallelPre and Post commands

HendriceH committed:
Fixed runParallelPre and Post commands

HendriceH committed:
Added kahip option for multilevel

greole committed:
Merge branch 'dev' into feat/custom_cmd

greole commented:
format!

lupeterm committed:
Format files

Co-authored-by: Gregor Olenik <greole@users.noreply.github.com>

lupeterm committed:
Format files

Co-authored-by: Gregor Olenik <greole@users.noreply.github.com>

greole commented on src/obr/cli.py:
```diff
@@ -219,7 +249,8 @@ def run(ctx: click.Context, **kwargs):
         return
 
-    if kwargs.get("operations") == "runParallelSolver":
+    # if kwargs.get("operations") == "runParallelSolver":
```
```suggestion
```

greole commented on src/obr/cli.py:
```diff
@@ -263,6 +295,12 @@ def run(ctx: click.Context, **kwargs):
+)
+@click.option(
+    "-e", "--env", is_flag=True, help="Shows required environment variables and exits."
+)
```
duplicate?
```suggestion
```

greole commented on src/obr/cli_util.py:
```diff
@@ -70,6 +70,8 @@ def check_cli_operations(
     """list available operations if none are specified or given the click option or an incorrect op is given"""
     if operations == ["generate"]:
         return True
+    if operations == ["simulate"]:
```
I am unsure whether to actually rename it to `execute` because simulate sounds like dry-run.
```suggestion
    if operations == ["execute"]:
```

greole commented on src/obr/signac_wrapper/operations.py:
```diff
@@ -677,13 +677,31 @@ def get_number_of_procs(job: Job) -> int:
 
 
+def get_tasks_per_node(job: Job) -> int:
+    """Deduces the number of tasks per node if variable is set. Else assumes HoreKa full not o 76"""
```
can we make that HoreKa agnostic?

greole commented on src/obr/signac_wrapper/operations.py:
```diff
@@ -699,35 +717,40 @@ def run_cmd_builder(job: Job, cmd_format: str, args: dict) -> str:
+    #    "path": job.path,
+    #    "timestamp": timestamp,
+    #    "np": get_number_of_procs(job),
+    # }
```
remove if uneeeded
```suggestion
```

greole reviewed (commented):
some quick comments

HendriceH committed:
Update submit.py

Changed "simulate" to "execute".

HendriceH committed:
Update submit.py

Changed "simulate" to "execute".

HendriceH committed:
Apply suggestions from code review

Cleanup.

Co-authored-by: Gregor Olenik <go@hpsim.de>

HendriceH committed:
Apply suggestions from code review

Cleanup.

Co-authored-by: Gregor Olenik <go@hpsim.de>

HendriceH committed:
Update operations.py

Changed from simulate to execute

HendriceH committed:
Update operations.py

Changed from simulate to execute

greole committed:
make get. query work outside of generator blocks

greole committed:
make get. query work outside of generator blocks

greole committed:
wip fix concurrent execution of pre and post cmd

greole committed:
wip fix concurrent execution of pre and post cmd

greole committed:
WIP: fixup pre and post calls

greole committed:
WIP: fixup pre and post calls

greole committed:
WIP, fixup create tree

greole committed:
WIP, fixup create tree

greole committed:
WIP, fixup uses

greole committed:
WIP, fixup uses

greole committed:
WIP, fixup uses

greole committed:
WIP, fixup uses

HendriceH committed:
Now using type in postprocess section of the YAML. Falls back to average if not given. Implemented diff_mean for calculation of time step size without intrumentation.

Now using type in postprocess section of the YAML. Falls back to average if not given. Implemented diff_mean for calculation of time step size without intrumentation.

HendriceH committed:
Now using type in postprocess section of the YAML. Falls back to average if not given. Implemented diff_mean for calculation of time step size without intrumentation.

Now using type in postprocess section of the YAML. Falls back to average if not given. Implemented diff_mean for calculation of time step size without intrumentation.

HendriceH committed:
Added diff_mean_skipX and diff_mean_useX types, where X is the writeInterval. With this one can filter the execution time or clocktime to use only the write time steps or skip the write timesteps and thus determine the writing overhead.

HendriceH committed:
Added diff_mean_skipX and diff_mean_useX types, where X is the writeInterval. With this one can filter the execution time or clocktime to use only the write time steps or skip the write timesteps and thus determine the writing overhead.

greole committed:
format

greole committed:
bump version

HendriceH committed:
runParallelSolver can now be used to run in serial (np=1)

HendriceH committed:
Guard tasksPerNode against not being set

HendriceH committed:
Fixed postprocess

HendriceH committed:
docs: map existing codebase

HendriceH committed:
Run pre_cmds/post_cmds verbatim around the solver in one submission

Since the submission script only contains 'obr run -o execute -j <id>' and
signac-flow executes every cmd operation in its own shell, separate pre/post
operations can never share shell state (e.g. a watchdog PID). Instead,
runParallelSolver (and runSerialSolver) now inline the pre_cmds/post_cmds
statepoint entries verbatim, newline-joined, into the single command string,
so lines like 'of_watchdog.sh --loop &', 'watchdog_pid=${!}' and
'kill $watchdog_pid' run in one shell in order.

- add get_raw_cmds/record_raw_cmds/wrap_raw_cmds helpers (no .format(), no
  wrapping - raw shell survives untouched, history provenance recorded)
- slim runParallelPre/runParallelPost down to standalone verbatim emitters,
  restore their has_pre_cmds/has_post_cmds preconditions, drop np/tpn
  directives and stale @pre.after orderings
- cli: 'obr run -o execute' takes the serial-safe path and comma-separated
  operations are no longer dropped
- tests for verbatim emission, ordering, inheritance; docs with the
  watchdog example and caveats
